internal/transport/http2_client.go

```diff
@@ -1179,14 +1181,139 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 		return
 	}
 
-	state := &decodeState{}
-	// Initialize isGRPC value to be !initialHeader, since if a gRPC Response-Headers has already been received, then it means that the peer is speaking gRPC and we are in gRPC mode.
-	state.data.isGRPC = !initialHeader
-	if err := state.decodeHeader(frame); err != nil {
+	// frame.Truncated is set to true when framer detects that the current header
+	// list size hits MaxHeaderListSize limit.
+	if frame.Truncated {
+		err := status.Error(codes.Internal, "peer header list size exceeded limit")
 		t.closeStream(s, err, true, http2.ErrCodeProtocol, status.Convert(err), nil, endStream)
 		return
 	}
 
+	var (
+		// If a gRPC Response-Headers has already been received, then it means that the peer is speaking gRPC and we are in gRPC mode.
+		isGRPC         = !initialHeader
+		mdata          = make(map[string][]string)
+		statusGen      *status.Status
+		grpcErr        error
+		httpErr        error
+		rawStatusCode  *int
+		httpStatus     *int
+		encoding       string
+		contentTypeErr string
+		rawStatusMsg   string
+	)
```

Comment:
Is it possible to eliminate some/all/most of these vars by doing the logic in-line?

E.g. instead of setting `isGRPC` and then checking it after the loop, move the contents of the `if isGRPC` to where we set it.

Reply:
I was able to remove the grpcErr and httpErr fields. I've tried to remove other fields but wasn't able to. Please let me know what you think!